xline/src/server/command.rs

```diff
@@ -37,39 +52,66 @@ pub(crate) struct KeyRange {
 impl KeyRange {
     /// Return if `KeyRange` is conflicted with another
     pub(crate) fn is_conflicted(&self, other: &Self) -> bool {
-        // All keys conflict with any key range
-        if self.end == ALL_KEYS || other.end == ALL_KEYS {
-            true
+        // s1 < s2 ?
+        if match (self.start_bound(), other.start_bound()) {
+            (Bound::Included(s1), Bound::Included(s2)) => {
+                if s1 == s2 {
+                    return true;
+                }
+                s1 < s2
+            }
+            (Bound::Included(_), Bound::Unbounded) => false,
+            (Bound::Unbounded, Bound::Included(_)) => true,
+            (Bound::Unbounded, Bound::Unbounded) => return true,
+            _ => unreachable!("KeyRange::start_bound() cannot be Excluded"),
+        } {
+            // s1 < s2
+            // s2 < e1 ?
+            match (other.start_bound(), self.end_bound()) {
+                (Bound::Included(s2), Bound::Included(e1)) => s2 <= e1,
+                (Bound::Included(s2), Bound::Excluded(e1)) => s2 < e1,
+                (Bound::Included(_), Bound::Unbounded) => true,
+                // if other.start_bound() is Unbounded, programe cannnot enter this branch
+                // KeyRange::start_bound() cannot be Excluded
+                _ => unreachable!("other.start_bound() should be Include"),
+            }
         } else {
-            match self.end.as_slice() {
-                ONE_KEY => match other.end.as_slice() {
-                    ONE_KEY => self.start == other.start,
-                    _ => (self.start >= other.start) && (self.start < other.end),
-                },
-                _ => match other.end.as_slice() {
-                    ONE_KEY => (other.start >= self.start) && (other.start < self.end),
-                    _ => (self.start < other.end) && (self.end > other.start),
-                },
+            // s2 < s1
+            // s1 < e2 ?
+            match (self.start_bound(), other.end_bound()) {
+                (Bound::Included(s1), Bound::Included(e2)) => s1 <= e2,
+                (Bound::Included(s1), Bound::Excluded(e2)) => s1 < e2,
+                (Bound::Included(_), Bound::Unbounded) => true,
+                // if self.start_bound() is Unbounded, programe cannnot enter this branch
+                // KeyRange::start_bound() cannot be Excluded
+                _ => unreachable!("self.start_bound() should be Include"),
             }
         }
     }
 
     /// Check if `KeyRange` contains a key
     pub(crate) fn contains_key(&self, key: &[u8]) -> bool {
-        match self.end.as_slice() {
-            ONE_KEY => self.start == key,
-            ALL_KEYS => true,
-            _ => (key >= self.start.as_slice()) && (key < self.end.as_slice()),
-        }
+        self.contains(key)
+    }
+
+    pub(crate) fn range_type(&self) -> RangeType {
```

Comment:
Add docs. Do you know why clippy doesn't catch this?

Reply:
fixed in #50